### PR TITLE
feat: skeleton support for OccurrenceAnalyzer

### PR DESCRIPTION
**Pull request by ChrBondi**

### CODE REVIEW

File: main/src/ca/uwaterloo/flix/language/phase/OccurrenceAnalyzer.scala [hunk 1]
```diff
@@ -0,0 +1,354 @@
+package ca.uwaterloo.flix.language.phase
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/OccurrenceAnalyzer.scala:1]: Remember to add copyright headers.

File: main/src/ca/uwaterloo/flix/language/phase/OccurrenceAnalyzer.scala [hunk 2]
```diff
@@ -0,0 +1,354 @@
...
+import ca.uwaterloo.flix.language.ast.LiftedAst.Expression
+import ca.uwaterloo.flix.language.ast.{LiftedAst, OccurrenceAst, Symbol}
+import ca.uwaterloo.flix.util.Validation
+import ca.uwaterloo.flix.util.Validation.ToSuccess
+import ca.uwaterloo.flix.language.ast.OccurrenceAst.Occur._
+import ca.uwaterloo.flix.language.ast.OccurrenceAst.Occur
+
+object OccurrenceAnalyzer {
+
+
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/OccurrenceAnalyzer.scala:14]: Extra newline.

File: main/src/ca/uwaterloo/flix/language/phase/OccurrenceAnalyzer.scala [hunk 3]
```diff
@@ -0,0 +1,354 @@
...
+  def run(root: LiftedAst.Root)(implicit flix: Flix): Validation[OccurrenceAst.Root, CompilationMessage] = flix.phase("OccurrenceAnalyzer") {
+    // TODO: Implement occurrence analysis.
+    val defs = root.defs.map {
+      case (sym, defn) =>
+        val fparams = defn.fparams.map { case LiftedAst.FormalParam(sym, mod, tpe, loc) => OccurrenceAst.FormalParam(sym, mod, tpe, loc) }
+        val (e, _) = visitExp(defn.exp)
+        sym -> OccurrenceAst.Def(defn.ann, defn.mod, defn.sym, fparams, e, defn.tpe, defn.loc)
+    }
+    val enums = root.enums.map {
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/OccurrenceAnalyzer.scala:23]: What is this computation for?

File: main/src/ca/uwaterloo/flix/language/phase/OccurrenceAnalyzer.scala [hunk 4]
```diff
@@ -0,0 +1,354 @@
...
+      case (sym, enum) =>
+        val cases = enum.cases.map {
+          case (tag, caze) =>
+            tag -> OccurrenceAst.Case(caze.sym, tag, caze.tpeDeprecated, caze.loc)
+        }
+        sym -> OccurrenceAst.Enum(enum.mod, enum.sym, cases, enum.tpeDeprecated, enum.loc)
+    }
+    // Reassemble the ast root.
+    val result = OccurrenceAst.Root(defs, enums, root.reachable, root.sources)
+    return result.toSuccess
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/OccurrenceAnalyzer.scala:33]: Return is implicit in Scala.

File: main/src/ca/uwaterloo/flix/language/phase/OccurrenceAnalyzer.scala [hunk 5]
```diff
@@ -0,0 +1,354 @@
...
+  }
+
+  def visitExp(exp0: LiftedAst.Expression): (OccurrenceAst.Expression, Map[Symbol.VarSym, OccurrenceAst.Occur]) = exp0 match {
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/OccurrenceAnalyzer.scala:36]: All functions other than run should be private and a have short description.

File: main/src/ca/uwaterloo/flix/language/phase/OccurrenceAnalyzer.scala [hunk 6]
```diff
@@ -0,0 +1,354 @@
...
+import ca.uwaterloo.flix.api.Flix
+import ca.uwaterloo.flix.language.CompilationMessage
+import ca.uwaterloo.flix.language.ast.LiftedAst.Expression
+import ca.uwaterloo.flix.language.ast.{LiftedAst, OccurrenceAst, Symbol}
+import ca.uwaterloo.flix.util.Validation
+import ca.uwaterloo.flix.util.Validation.ToSuccess
+import ca.uwaterloo.flix.language.ast.OccurrenceAst.Occur._
+import ca.uwaterloo.flix.language.ast.OccurrenceAst.Occur
+
+object OccurrenceAnalyzer {
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/OccurrenceAnalyzer.scala:12]: Add a few sentences that explain what this phase does.

File: main/src/ca/uwaterloo/flix/language/phase/OccurrenceAnalyzer.scala [hunk 7]
```diff
@@ -0,0 +1,354 @@
...
+        val (e, o3) = visitExp(r.exp)
+        rs = rs :+ OccurrenceAst.SelectChannelRule(r.sym, c, e)
+        o1 = merge(o1, merge(o2, o3))
+      }
+      val (d, o4) = default match {
+        case Some(exp) =>
+          val (d, o5) = visitExp(exp)
+          (Some(d), merge(o5, o1))
+        case None => (None, o1)
+      }
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/OccurrenceAnalyzer.scala:300]: Introduce a helper function for working with options?

File: main/src/ca/uwaterloo/flix/language/phase/OccurrenceAnalyzer.scala [hunk 8]
```diff
@@ -0,0 +1,354 @@
...
+    args.foldLeft((List[OccurrenceAst.Expression](), Map[Symbol.VarSym, OccurrenceAst.Occur]()))((acc, arg) => {
+      val (e, o) = visitExp(arg)
+      val m = merge(o, acc._2)
+      (acc._1 :+ e, m)
+    })
+  }
+
+  def merge(m1: Map[Symbol.VarSym, Occur], m2: Map[Symbol.VarSym, Occur]): Map[Symbol.VarSym, Occur] = {
+
+    def combine(o1: Occur, o2: Occur): Occur = (o1, o2) match {
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/OccurrenceAnalyzer.scala:332]: Move to top-level and add comments.

File: main/src/ca/uwaterloo/flix/language/phase/OccurrenceAnalyzer.scala [hunk 9]
```diff
@@ -0,0 +1,354 @@
...
+      case (_, Dead) => o1
+      case _ => Many
+    }
+
+    var m3 = m1.map {
+      case (sym, o1) =>
+        val opt = m2.get(sym)
+        opt match {
+          case Some(o2) => (sym, combine(o1, o2))
+          case None => (sym, o1)
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/OccurrenceAnalyzer.scala:343]: None cases are by convention always first.

File: main/src/ca/uwaterloo/flix/language/phase/OccurrenceAnalyzer.scala [hunk 10]
```diff
@@ -0,0 +1,354 @@
...
+
+  def visitExps(args: List[LiftedAst.Expression]): (List[OccurrenceAst.Expression], Map[Symbol.VarSym, Occur]) = {
+    args.foldLeft((List[OccurrenceAst.Expression](), Map[Symbol.VarSym, OccurrenceAst.Occur]()))((acc, arg) => {
+      val (e, o) = visitExp(arg)
+      val m = merge(o, acc._2)
+      (acc._1 :+ e, m)
+    })
+  }
+
+  def merge(m1: Map[Symbol.VarSym, Occur], m2: Map[Symbol.VarSym, Occur]): Map[Symbol.VarSym, Occur] = {
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/OccurrenceAnalyzer.scala:330]: I would rename this to combineAll

File: main/src/ca/uwaterloo/flix/language/phase/OccurrenceAnalyzer.scala [hunk 11]
```diff
@@ -0,0 +1,354 @@
...
+    case Expression.Closure(sym, freeVars, tpe, loc) =>
+      val freeVars1 = freeVars.map {
+        case LiftedAst.FreeVar(sym, tpe) => OccurrenceAst.FreeVar(sym, tpe)
+      }
+      (OccurrenceAst.Expression.Closure(sym, freeVars1, tpe, loc), Map.empty)
+
+    case Expression.ApplyClo(exp, args, tpe, loc) =>
+      val (e, o1) = visitExp(exp)
+      val (as, o2) = visitExps(args)
+      val m = merge(o1, o2)
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/OccurrenceAnalyzer.scala:74]: Here you use the name `m`, but in other places you use `o3` and in other places `bs`. Find one convention and stick to it :)

File: main/src/ca/uwaterloo/flix/language/phase/OccurrenceAnalyzer.scala (hunk 3 above)
Comment (ChrBondi) [reply main/src/ca/uwaterloo/flix/language/phase/OccurrenceAnalyzer.scala:23]: Added comment explaining that it transforms LiftedAst Defs to OccurrenceAst Defs

File: main/src/ca/uwaterloo/flix/language/phase/OccurrenceAnalyzer.scala [hunk 12]
```diff
@@ -0,0 +1,396 @@
+/*
+ * Copyright 2022 Magnus Madsen, Anna Krogh, Patrick Lundvig, Christian Bonde
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/OccurrenceAnalyzer.scala:2]: You don't have to add my name, but whatever :)

File: main/src/ca/uwaterloo/flix/language/phase/OccurrenceAnalyzer.scala [hunk 13]
```diff
@@ -0,0 +1,396 @@
...
+import ca.uwaterloo.flix.language.ast.OccurrenceAst.Occur
+import ca.uwaterloo.flix.language.ast.OccurrenceAst.Occur._
+import ca.uwaterloo.flix.language.ast.{LiftedAst, OccurrenceAst, Symbol}
+import ca.uwaterloo.flix.util.Validation
+import ca.uwaterloo.flix.util.Validation.ToSuccess
+
+/**
+ * The occurrence analyzer collects information on variable usage
+ */
+object OccurrenceAnalyzer {
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/OccurrenceAnalyzer.scala:31]: We can call it OccurrenceAnalyzer for now, maybe later we can find a better name.

File: main/src/ca/uwaterloo/flix/language/phase/OccurrenceAnalyzer.scala [hunk 14]
```diff
@@ -0,0 +1,396 @@
...
+import ca.uwaterloo.flix.language.CompilationMessage
+import ca.uwaterloo.flix.language.ast.LiftedAst.Expression
+import ca.uwaterloo.flix.language.ast.OccurrenceAst.Occur
+import ca.uwaterloo.flix.language.ast.OccurrenceAst.Occur._
+import ca.uwaterloo.flix.language.ast.{LiftedAst, OccurrenceAst, Symbol}
+import ca.uwaterloo.flix.util.Validation
+import ca.uwaterloo.flix.util.Validation.ToSuccess
+
+/**
+ * The occurrence analyzer collects information on variable usage
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/OccurrenceAnalyzer.scala:29]: Is this what passes for good documentation in industry? 

image

File: main/src/ca/uwaterloo/flix/language/phase/OccurrenceAnalyzer.scala [hunk 15]
```diff
@@ -0,0 +1,396 @@
...
+      val o2 = combineAll(o1, acc._2)
+      (acc._1 :+ e, o2)
+    })
+  }
+
+  /**
+   * Combines the 2 maps `m1` and `m2` of the type (Symbol -> Occur) into a single map of same type using the function `combine`.
+   */
+  private def combineAll(m1: Map[Symbol.VarSym, Occur], m2: Map[Symbol.VarSym, Occur]): Map[Symbol.VarSym, Occur]
+  =
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/OccurrenceAnalyzer.scala:362]: Whats with the = on a new line? :)

File: main/src/ca/uwaterloo/flix/language/phase/OccurrenceAnalyzer.scala [hunk 16]
```diff
@@ -0,0 +1,396 @@
...
+  private def combine(o1: Occur, o2: Occur): Occur = (o1, o2) match {
+    case (Dead, _) => o2
+    case (_, Dead) => o1
+    case _ => Many
+  }
+
+  /**
+   * Matches on `opt`. If None returns `default`, else apply `f` on value in Some(B)
+   */
+  private def resolveOption[A, B](opt: Option[B], default: A, f: B => A): A = {
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/OccurrenceAnalyzer.scala:388]: OK, and now you can realize it is a function in the standard library. In Scala its `.map(...).getOrElse(...)`.

File: main/src/ca/uwaterloo/flix/language/phase/OccurrenceAnalyzer.scala [hunk 17]
```diff
@@ -0,0 +1,396 @@
...
+      case (sym, o1) =>
+        val o3 = resolveOption(m2.get(sym), o1, combine(o1, _))
+        (sym, o3)
+    }
+
+    m2.foreach {
+      case (sym, o) => if (!m1.contains(sym)) m3 += (sym -> o)
+    }
+    m3
+  }
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/OccurrenceAnalyzer.scala:374]: How about: 
```scala
(m1.keys ++ m2.keys).foldLeft(Map.empty){
  case (acc, k) => acc + (k -> mergeOpt(m1.get(k), m2.get(k)))
})

def mergeOpt(o1: Option[Occur], o2: Option[Occur]): Occur = ...
```
(Not tested)

**Review by magnus-madsen — COMMENTED** [review]
Very good!

I added a few last comments that you can fix and think about and then we can merge it!

**Comment by magnus-madsen** [comment]
Merged!

**Pull request merged**